sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1396 @@
+
+Parameters:
+
+* `graph` - Modifiable graph object to update graph node inputs & outputs with.
+  This graph must have the same topology as the original graph used on
+  executable graph creation.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if the topology of `graph` is
+  not the same as the existing graph topology, or if the nodes were not added in
+  the same order.
+|===
+
+=== Queue Class Modifications
+
+:queue-class: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:interface.queue.class
+
+This extension modifies the {queue-class}[SYCL queue class] such that
+<<queue-state, state>> is introduced to queue objects, allowing an instance to be
+put into a mode where command-groups are recorded to a graph rather than
+submitted immediately for execution.
+
+<<new-queue-member-functions, Three new member functions>> are also added to the
+`sycl::queue` class in this extension as queue shortcuts for `handler::graph()`.
+
+==== Queue State
+
+:queue-info-table: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.queue.info
+
+The `sycl::queue` object can be in either of two states. The default
+`queue_state::executing` state is where the queue has its normal semantics of
+submitted command-groups being immediately scheduled for asynchronous execution.
+
+The alternative `queue_state::recording` state is used for graph construction.
+Instead of being scheduled for execution, command-groups submitted to the queue
+are recorded to a graph object as new nodes for each submission. After recording
+has finished and the queue returns to the executing state, the recorded commands are
+not then executed, they are transparent to any following queue operations.
+
+.Queue State Diagram
+[source, mermaid]
+....
+graph LR
+    Executing -->|Begin Recording| Recording
+    Recording -->|End Recording| Executing
+....
+
+The state of a queue can be queried with `queue::get_info` using template
+parameter `info::queue::state`. The following entry is added to the
+{queue-info-table}[queue info table] to define this query:
+
+Table {counter: tableNumber}. Queue info query
+[cols="2a,a,a"]
+|===
+| Queue Descriptors | Return Type | Description
+
+| `info::queue::state`
+| `ext::oneapi::experimental::queue_state`
+| Returns the state of the queue
+
+|===
+
+Events returned from queue submissions when a queue is in the recording state
+may only be used as parameters to `handler::depends_on()` or as dependent
+events for queue shortcuts like `queue::parallel_for()` for submissions which
+are being recorded to the same modifiable `command_graph`. These events have
+status `info::event_command_status::complete`. The event status of an event
+returned from an executable graph submission will have
+`info::event_command_status::running` once any command group node starts
+executing on a device, and status `info::event_command_status::complete`
+once all the nodes have finished execution.
```

Comment:
> The behavior you describe here is weird because the status goes "backwards" from complete to running

That's not the behaviour i'm trying to describe, these two event scenarios described here:
1) For an event returned from graph a submission, the status goes from "complete" to "running"
2) For an event from returned from a record & replay queue submissions, the status always stays at "complete".

For banning `get_info<info::event::command_execution_status>()` would that be both situations? Or just scenario 1)? For scenario 1) this makes me think we should ban `get_profiling_info()` on those events too

Reply:
I meant scenario (2).  It seems wrong to return "complete" for this event because no command has executed.  An application could reasonably assume that it can read the result of a command after the event status is "complete".  In the record & replay case, this would lead to wrong results because the command didn't actually execute.

Reply:
Making checking the event status or profiling info an exception in https://github.com/reble/llvm/pull/132